src/sentry/services/hybrid_cloud/user.py

```diff
@@ -0,0 +1,147 @@
+from abc import abstractmethod
+from dataclasses import dataclass
+from inspect import signature
+from typing import Any, Iterable, List, Optional, Sequence, Union
+
+from sentry.models.group import Group
+from sentry.models.organizationmember import OrganizationMember
+from sentry.models.user import User
+from sentry.services.hybrid_cloud import (
+    CreateStubFromBase,
+    InterfaceWithLifecycle,
+    silo_mode_delegation,
+)
+from sentry.silo import SiloMode
+
+
+@dataclass(frozen=True)
+class APIUser:
+    id: int = -1
+    pk: int = -1
+    name: str = ""
+    email: str = ""
+    username: str = ""
+    actor_id: int = -1
+
+    # TODO: Extract these "view" methods to BaseUser class?
+    # BaseUser class would also be useful for catching `isinstance(foo, User)` checks
+    # which may no longer mean what they used to mean
+    def class_name(self):
+        """Necessary for integrations that introspect on user objects"""
+        return "User"
+
+    def get_display_name(self):
+        return self.name or self.email or self.username
+
+    # TODO: Extract to base dataclass?
+    @classmethod
+    def from_orm(cls, orm: Any):
```

Comment:
There will come a time I think when this logic has to go and be replaced with explicit attribute packing.  The problem is that not all attributes will carry the same typing by the end, a good example being relationships, but some others as well.

Reply:
I'm not sure I agree.
Relationship attributes are named without the `_id` suffix. E.g. `.actor` is the relationship, `.actor_id` is the data value. So we just copy the data values, [e.g.](https://github.com/getsentry/sentry/pull/40614/files/0441a3c3579dbf742bd6f03fdee699db77bdea84#diff-5ee4410f00e99e25c0534d979882b2211d87337651d4083076f9af13340179caR24)

Reply:
You mentioned you wanted to avoid this becoming "too rpc-y".  As a result, I've been refactoring in such a way that many domain relations come "packed" into an object, which is going to be much more performant and correct in the end.  In that case, something like this cannot possibly work:

https://github.com/getsentry/sentry/pull/40772/files#diff-4c7c9fddec7c4378c1d12c024acfd6901b6c74750350cda751ef79970bbd7d70R90

Reply:
Ah yes, this is definitely true.